### PR TITLE
Fix folder moving in file system dock

### PR DESCRIPTION
**Pull request by Rindbee**
Include files in the folder that will be moved/renamed when looking up the file owner. 

Fixed issue when moving/renaming directories, see https://github.com/godotengine/godot/issues/77289#issuecomment-1722027314.

Rename the function/variable names to make more sense.

### CODE REVIEW

File: editor/filesystem_dock.cpp [hunk 1]
```diff
@@ -1363,15 +1363,15 @@ void FileSystemDock::_get_all_items_in_dir(EditorFileSystemDirectory *p_efsd, Ve
...
-		_find_remaps(p_efsd->get_subdir(i), r_renames, r_to_remaps);
+		_find_file_owners(p_efsd->get_subdir(i), p_renames, r_file_owners);
 	}
 	for (int i = 0; i < p_efsd->get_file_count(); i++) {
 		Vector<String> deps = p_efsd->get_file_deps(i);
 		for (int j = 0; j < deps.size(); j++) {
-			if (r_renames.has(deps[j])) {
-				r_to_remaps.push_back(p_efsd->get_file_path(i));
+			if (p_renames.has(deps[j])) {
+				r_file_owners.push_back(p_efsd->get_file_path(i));
```
Comment (PinkWolfCoder) [inline editor/filesystem_dock.cpp:1374]:
```suggestion
				r_file_owners.emplace_back(p_efsd->get_file_path(i));
```

File: editor/filesystem_dock.cpp [hunk 2]
```diff
@@ -1560,22 +1560,22 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
+void FileSystemDock::_update_dependencies_after_move(const HashMap<String, String> &p_renames, const Vector<String> &p_file_owners) const {
 	// The following code assumes that the following holds:
 	// 1) EditorFileSystem contains the old paths/folder structure from before the rename/move.
 	// 2) ResourceLoader can use the new paths without needing to call rescan.
 	List<String> scenes_to_reload;
-	for (int i = 0; i < p_remaps.size(); ++i) {
+	for (int i = 0; i < p_file_owners.size(); ++i) {
 		// Because we haven't called a rescan yet the found remap might still be an old path itself.
-		String file = p_renames.has(p_remaps[i]) ? p_renames[p_remaps[i]] : p_remaps[i];
+		String file = p_renames.has(p_file_owners[i]) ? p_renames[p_file_owners[i]] : p_file_owners[i];
```
Comment (PinkWolfCoder) [inline editor/filesystem_dock.cpp:1570]: I think this one can be const

File: editor/filesystem_dock.cpp [hunk 3]
```diff
@@ -1560,22 +1560,22 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
 		print_verbose("Remapping dependencies for: " + file);
 		Error err = ResourceLoader::rename_dependencies(file, p_renames);
```
Comment (PinkWolfCoder) [inline editor/filesystem_dock.cpp:1572]: err could be const

File: editor/filesystem_dock.cpp [hunk 4]
```diff
@@ -1560,22 +1560,22 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
 		if (err == OK) {
 			if (ResourceLoader::get_resource_type(file) == "PackedScene") {
 				scenes_to_reload.push_back(file);
```
Comment (PinkWolfCoder) [inline editor/filesystem_dock.cpp:1575]:
```suggestion
				scenes_to_reload. emplace_back(file);
```

File: editor/filesystem_dock.cpp [hunk 5]
```diff
@@ -1949,20 +1947,43 @@ void FileSystemDock::_move_operation_confirm(const String &p_to_path, bool p_cop
...
+				for (int j = 0; j < current_folder->get_file_count(); j++) {
+					String file_path = current_folder->get_file_path(j);
+					renamed_files.push_back(file_path);
+					ResourceUID::ID uid = ResourceLoader::get_resource_uid(file_path);
+					if (uid != ResourceUID::INVALID_ID) {
+						r_uids[file_path] = uid;
+					}
+				}
+				for (int j = 0; j < current_folder->get_subdir_count(); j++) {
+					folders.push_back(current_folder->get_subdir(j));
```
Comment (PinkWolfCoder) [inline editor/filesystem_dock.cpp:1974]: maybe we can use emplace_back here

File: editor/filesystem_dock.cpp (hunk 1 above)
Comment (Rindbee) [reply editor/filesystem_dock.cpp:1374]: Some things are different in Godot. See  [Disallowed features](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#disallowed-features).

Comment (AThousandShips) [reply editor/filesystem_dock.cpp:1374]: `Vector` does not have `emplace_back`

Comment (PinkWolfCoder) [reply editor/filesystem_dock.cpp:1374]: Cool thanks!

Comment (AThousandShips) [reply editor/filesystem_dock.cpp:1374]: Reviews are much appreciated but make sure you understand the codebase and the code you are reviewing

Comment (PinkWolfCoder) [reply editor/filesystem_dock.cpp:1374]: Will do!

**Comment by KoBeWi** [comment]
It's still possible to break dependencies if you rename a folder (tested with inherited scenes).

Also doing specific steps, you can cause a crash:
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.dev.custom_build (4c3dc26367518e006f8555c12f5d2df0b8a28192)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] CowData<EditorFileSystemDirectory::FileInfo *>::_get_size (C:\godot_source\core\templates\cowdata.h:81)
[1] CowData<EditorFileSystemDirectory::FileInfo *>::size (C:\godot_source\core\templates\cowdata.h:134)
[2] Vector<EditorFileSystemDirectory::FileInfo *>::size (C:\godot_source\core\templates\vector.h:93)
[3] EditorFileSystemDirectory::get_file_count (C:\godot_source\editor\editor_file_system.cpp:91)
[4] FileSystemDock::_before_move (C:\godot_source\editor\filesystem_dock.cpp:1967)
[5] FileSystemDock::_rename_operation_confirm (C:\godot_source\editor\filesystem_dock.cpp:1780)
[6] call_with_variant_args_helper<FileSystemDock> (C:\godot_source\core\variant\binder_common.h:308)
[7] call_with_variant_args<FileSystemDock> (C:\godot_source\core\variant\binder_common.h:418)
[8] CallableCustomMethodPointer<FileSystemDock>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[9] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[10] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1118)
[11] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3596)
[12] Object::emit_signal<> (C:\godot_source\core\object\object.h:920)
[13] Tree::item_edited (C:\godot_source\scene\gui\tree.cpp:4342)
[14] Tree::_line_editor_submit (C:\godot_source\scene\gui\tree.cpp:3131)
[15] call_with_variant_args_helper<Tree,String,0> (C:\godot_source\core\variant\binder_common.h:308)
[16] call_with_variant_args<Tree,String> (C:\godot_source\core\variant\binder_common.h:418)
[17] CallableCustomMethodPointer<Tree,String>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[18] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[19] Object::emit_signalp (C:\godot_source\core\object\object.cpp:1118)
[20] Node::emit_signalp (C:\godot_source\scene\main\node.cpp:3596)
[21] Object::emit_signal<String> (C:\godot_source\core\object\object.h:920)
[22] LineEdit::gui_input (C:\godot_source\scene\gui\line_edit.cpp:480)
[23] Control::_call_gui_input (C:\godot_source\scene\gui\control.cpp:1810)
[24] Viewport::_gui_input_event (C:\godot_source\scene\main\viewport.cpp:2236)
[25] Viewport::push_input (C:\godot_source\scene\main\viewport.cpp:3186)
[26] Window::_window_input (C:\godot_source\scene\main\window.cpp:1549)
[27] call_with_variant_args_helper<Window,Ref<InputEvent> const &,0> (C:\godot_source\core\variant\binder_common.h:303)
[28] call_with_variant_args<Window,Ref<InputEvent> const &> (C:\godot_source\core\variant\binder_common.h:418)
[29] CallableCustomMethodPointer<Window,Ref<InputEvent> const &>::call (C:\godot_source\core\object\callable_method_pointer.h:105)
[30] Callable::callp (C:\godot_source\core\variant\callable.cpp:51)
[31] DisplayServerWindows::_dispatch_input_event (C:\godot_source\platform\windows\display_server_windows.cpp:2628)
[32] DisplayServerWindows::_dispatch_input_events (C:\godot_source\platform\windows\display_server_windows.cpp:2605)
[33] Input::_parse_input_event_impl (C:\godot_source\core\input\input.cpp:751)
[34] Input::flush_buffered_events (C:\godot_source\core\input\input.cpp:1013)
[35] DisplayServerWindows::process_events (C:\godot_source\platform\windows\display_server_windows.cpp:2317)
[36] OS_Windows::run (C:\godot_source\platform\windows\os_windows.cpp:1474)
[37] widechar_main (C:\godot_source\platform\windows\godot_windows.cpp:182)
[38] _main (C:\godot_source\platform\windows\godot_windows.cpp:204)
[39] main (C:\godot_source\platform\windows\godot_windows.cpp:218)
[40] WinMain (C:\godot_source\platform\windows\godot_windows.cpp:232)
[41] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[42] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

https://github.com/godotengine/godot/assets/2223172/3a66bd35-8996-4690-b036-3bc1cf761e2d

(B inherits A)

**Comment by Rindbee** [comment]
> It's still possible to break dependencies if you rename a folder (tested with inherited scenes).
> Also doing specific steps, you can cause a crash:

`to_move` was not used when renaming.
Has been fixed, and rebased.

File: editor/filesystem_dock.cpp [hunk 6]
```diff
@@ -1562,22 +1562,22 @@ void FileSystemDock::_update_resource_paths_after_move(const HashMap<String, Str
...
 }
 
-void FileSystemDock::_update_dependencies_after_move(const HashMap<String, String> &p_renames, const Vector<String> &p_remaps) const {
+void FileSystemDock::_update_dependencies_after_move(const HashMap<String, String> &p_renames, const Vector<String> &p_file_owners) const {
 	// The following code assumes that the following holds:
 	// 1) EditorFileSystem contains the old paths/folder structure from before the rename/move.
 	// 2) ResourceLoader can use the new paths without needing to call rescan.
-	List<String> scenes_to_reload;
-	for (int i = 0; i < p_remaps.size(); ++i) {
+	Vector<String> scenes_to_reload;
```
Comment (KoBeWi) [inline editor/filesystem_dock.cpp:1569]: Why did you change it to Vector? A list is better in this case.

**Review by KoBeWi — APPROVED** [review]

File: editor/filesystem_dock.cpp (hunk 6 above)
Comment (Rindbee) [reply editor/filesystem_dock.cpp:1569]: In the "Note" below [Disallowed features](https://docs.godotengine.org/en/stable/contributing/development/cpp_usage_guidelines.html#disallowed-features), `Vector` seems to be more recommended. I have always used `List` as a queue.

Comment (KoBeWi) [reply editor/filesystem_dock.cpp:1569]: I'd expect list to perform better in case where you only append and iterate things. You don't know its size beforehand and don't need random access. But maybe Vector is implemented in a way that makes it more efficient here, idk. It doesn't really matter anyway, unless you have a thousand opened scenes.

Comment (Rindbee) [reply editor/filesystem_dock.cpp:1569]: Do I need to change it back?

Comment (KoBeWi) [reply editor/filesystem_dock.cpp]: You can leave it. Maybe someone more knowledgeable will give feedback.

**Comment by akien-mga** [comment]
Thanks!
